README
        
	   ===========================================
	   Vida - Visual Interactive Datapipe

	   no (c) 2002 - Embyte & Lesion - README FILE
	   =========================================== 
           

INTRO:
Datapipe is a pipe between two sockets like a "network pipeline".

From man bash:
A pipeline is a sequence of one or more commands separated by the character |. 
The format for a pipeline is : command1 | command2 
The standard output of command1 is connected to the standard input of command2.

Datapipe makes a pipe between a listen port on the machine it's being run on
and a port on the remote machine. For example if we create a datapipe like
this: 25->mail.tin.it:25 all the incoming smtp request to machine pipe it's
being run on will be redirect to mail.tin.it (and so telnetting to port 25 
on localhost is the same as telnetting to port 25 of tin smtp server). 

Datapipe can be used for a variety of purpose: 
redirect connection which require ident answer (sendmail or irc for example) 
so that identd show username of datapipe PID process;
redirect call for a daemon firewalled (for example if ssh is firewalled we
can create a datapipe on a non-firewalled port like : 5000->localhost:22);
using the future features dns-spoofing and arp-poisoning to redirect request
to own machine to sniff/hijacking/kill/log connection. 
This is the scenario:
normal connection:      victim user <--- telnet protocol ---> 23:server
redirected connection:  victim user <----> 23:attakker <----> 23:server
Attakker spoof dns-answer or make arp-poisoning (on switched LAN) and makes
unaware victim to connect to himself, then use a datapipe to redirect data
to server. Now attakker can see all traffic between victim and server, log it
or push his data into connection :-)
The same thing is possible using a html or javascript code (for example a
irc javascript client, like those present on web page of your favourite irc
server) so that victim girl think to join her favorite server, but connect
to attakker PC who bounce (or read/hijack) her private messages :-)    

MINI-DESCRIPTION:
Vida is a multi-datapipe handler, written in C with ncurses library
for unix and unix-like OS.

FEATURES:
- visual interface built with ncurses library
- multiple datapipes
- multiple connections for each datapipe
- computes the total of sent and received bytes
- sniffing a routed connection (monodirectional, bidirectional and to colors)
- simple hijacking of a routed connection (in both  directions)
- dns-hijacking of a dns-spoofed connection over switched LAN
- some logging formats
- * some features are not yet implementated *
- * some features will be implementated in the future release *

HELP:
Read manpage (type 'man help' form shell) or select 'Help' item from Vida
main menu

INSTALL:
To install Vida you must:
0. Check and install dependence (ncurses, libnet and libpcap library).
   Retrive theme here:
   libncurses    <http://www.gnu.org/software/ncurses/ncurses.html>
   libnet        <http://www.packetfactory.net/libnet/>
   libpcap       <http://www.tcpdump.org/#current>			
1. Extract tar.gz archive : 'tar xvfz vida-<version>.tar.gz'
2. Enter 'vida' directory : 'cd vida'
3. Read this README : 'less README' or 'more README'
4. Build Vida binary: 'make'
5. (Optional) Install Vida : 'make install'
   This command copy vida binary-file to /usr/bin and
   vida man-page (and online help) to /usr/man/man1
6. Run Vida with 'vida' or just './vida' if you have jumped previous point.
   
UNINSTALL:
Type 'make clean' or 'make install' from src directory (vida)

AVAILABILITY:
Vida project web site is at :
  http://vidatapipe.sourceforge.net
  http://freshmeat.net/projects/vida
Vida current version is downloadable at
  http://vidatapipe.sourceforge.net/vida-current.tar.gz
  
LICENSE and CHANGELOG:
Read 'LICENSE' file 
Read 'CHANGELOG' file for english version and 'CHANGELOG_it' for italian

GREETING:
I (embyte) owe a special thank-you to my family and my girlfriend having
bear myself, and a great regard to all acari and friends :* 


CHANGELOG
vida-0.8devel: (middle of June 2003):
  - Finally has been added DNS Spoofing support. This is in BETA. 
  - Code cleaned
  - Add some english comment. It so much remains job to make.
  - Info() has been changed
  - indent use -kr flags but Embyte prefers jed :-)
  - Changed mvwaddstr with mvwprintw
  
vida-0.7.1:
  - Vida can now print packages that matched a given string (grep over
    TCP payload)
  - I have arranged manpage
  
vida-0.7:
  - datapipe password support - now you can use vida as an autenticated
    wingate, bouncer or proxy to hide your IP. I haven't implemented a 
    multy-account system but only a single password autentication which is much
    sufficient for a datapipe 
  - README file has being written (I think lesion has just included it the
    0.6 package, but it will be insert officialy from this release on)
  - correct some mini-bug
    
vida-0.6:
  - added file-logging support (use 'l' on the choosed connection to
    enable and disable) (vida creates a file IPrequest-IPbounced:PORT).
  - man page completed
  - vida is now on freshmeat.net
    (http://freshmeat.net/projects/vida/?topic_id=152%2C150%2C43)
  - has being insered a pseudo LICENSE file

vida-0.5 (a month later...):
  - I've changed (a few) the header files setup and reversed the history file
    (last modifies before) now named CHANGELOG;
  - there's a sketch of site ripped from paranoy (index.php);
  - the Makefile is filled :). "make" to compile, "make install"
    and "uninstall" are ok, and there's also "make indent" to indent the code
  - if a connection happens meanwhile the connection list is open it will be refreshed
    instantly
  - now the update of the amount of sent/received data is for all the connections
    and not only for the selected one :)
  - fixed some bugs of displaying in datapipe()
  - the printing in wdata now works!! there are 4 switchable modes pressing
    enter on the chose connection or with special keys.
    The first prints incoming data from the client, the second one prints
    the incoming data from server, the third mode prints both, and the fourth
    mode is the same as the third except it changes the colour for each provenience.
    Every mode change is evidenced in wdata.
    Mode 1 can be selected with 'c' (client) key, mode 2 with 's' (server),
    mode 3 uses 'b' (both) key and mode 4 with 'k' (c letter was busy and
    hence kolor (such as pine :) ))
  - when I pick out a connection to bring it in 'sniffing' mode, the status flags
    of the other are zeroed with statzero().
  - some variables changes for the simple hijacking
  - sourceforge registration requested
  - sketch of man page
  - "make install" install also man pages

vida-0.4:
  I've added "simple hijacking" feature allowing, with the 'j' key press on the
  selected connection, to send a data string straight to either a direction or the
  other (simple implementation...but very interesting for studying how the nets work
  and also for ludus. (irc hijacking :pP)).
  - the "help" item shows the HELP file scrollable up and down (ripped and fitted
    from paranox.sourceforge.net)
  - i've added the banner onto each source.
  - i've written (all alone) this HISTORY that it will evolve in a much schematic
    way.


vida-0.3:
  Starting with this release vida is divided in separate files for a better maintenance
  and optimisation of the code (to compile simply type make, all is going to be read
  from the Makefile). lesion has fixed some bugs related to the delete of the connections,
  and moreover:
  - lesion is working to add the dns request interception to spoofing them and
    redirect the traffic throughout vida. (one part of the code was ripped
    (see dnshijacking.c))
  - I think that the vida's project web site is up and running on sourceforge (it will
    be available also on freshmeat)


vida-0.2.3:
  A lot of bugs were fixed and some pretty features were added or improved:
  - deleting of connections or datapipes (available with 'canc' or 'd' key)
    is now almost perfect; there is only a problem in the accept() of datapipe()
    during the pthread_cancel() function call). The thread ain't be deleted and
    the bound port remains in listen mode.
  - It's possible to print and sniff the traffic throughout the datapipe and
    in the selected connection (with the 'enter' key) (this feature was removed
    between 0.2 and 0.2.3 releases). However there is basically a problem: it works
    only under X and hence vida lose a big slice of its charming, thus we decided
    to remove this feature (available only in the 0.2.3 release) in the following
    versions until the release of a full working one. (lesion is thinking about the
    raw use of the terminal from LINES/2 to the bottom).
  - Near each datatipe you can see in real time the number of the connections that it's
    going to handle
  - Near each connection (in the datapipe's menu) is it possible to see in real time the
    amount of transferred bytes (in both directions)
    [remember that a single word like 'user(enter)' takes 5/6 bytes (4 of user +
    1 for the \n and another 1 for the \r if the client is, for example, telnet)].



vida-0.2:
  At this time lesion join the game. He improves very much the vida project.
  A lot of send-and-receive (code and emails) between us allow the birth the
  0.2 release, totally rewritten using the ncurses library (from its man page:
  CRT screen handling and optimisation package). Vida now supports more datapipes
  at the same time and as before a lot of connections for each (It was developed
  with the aid of two nested threads, the outer one handles the datapipes (see
  datapipe() func.) and the inner one routes the whole traffic to the single
  connections (see handler() func.).


vida-0.1:
  I've fixed to things:

include.h
/*____________________________________________*\
*                                              *
*                                              *
*                -= vida =-                    *
*       Visual Interactive Datapipe            *
*          no (c) 2002/03-2003/05              *
*                  v.0.8.0                     *
*                                              *
*    coded by                                  *
*    embyte -> embyte@madlab.it                *
*    lesion -> lesion@autisti.org              *
*                                              *
*    less README for features and more info    *
*                                              *
*    --------------------------------------    *
*                                              *
*    include.h:                                *
*      -declare includes                       *
*      -define some vars                       *
*      -declare some needed structs            *
*      -declare global vars                    *
*      -do prototype of functions              *
*                                              *
*                                              *
\*____________________________________________*/


/* Very good libraries which make ports very easy ;)  */

#include <libnet.h>
#include <pcap.h>
#include <pthread.h>
#include <ncurses.h>

#define VERSION "0.8devel"
#define MAXLINE 1029

#define CHECKPWD(n) (dp[n].check=='Y' || dp[n].check=='y')
struct redirect {

    int localport;
    int remoteport;
    char *host;

    // numero di connessioni di questo datapipe
    // number of connection of this datapipe
    ushort cn;

    // connessioni di questo datapipe (hosts)
    // connection of this datapipe (hosts)
    char *ch[10];

    // connessioni di questo datapipe (ports)
    // connection of this datapipe (ports)

    ushort cp[10];

    // byte ricevuti
    // received byted
    long int rb[10];

    // byte trasmessi
    // sent bytes
    long int tb[10];

    //socket descriptor
    int from_sd[10];
    int dest_sd[10];
    struct sockaddr_in sout;

    /*
     * datapipe connection status value: 
     0- print "client" data 
     1- print "server" data 
     2- print all data 
     3- print all data in different color 
     4- print "matched" data
     ---- 
     5-hijacking (keep client)
     6-hijacking (keep server) 
     7-hijacking (keep both) (skleroooooooooooo)
     8-closing this connection
     */
    int status[10];
    int logfile[10];
    //datapipe status
    int stat;
    FILE *fout;
	char mstring[50];
    char check;
    char pwd[32];
} dp[10];

//number of used datapipe
int ndp;



pthread_t dp_thread[10], h;

// windows
WINDOW *wmenu, *wlist, *wdata;

//this shows where is focus
// 0 = wmenu, 1 = wlist (list datapipe), 2 = wlist (list connection)
int win;

// mpos shows cursor's position in menu

// dpos shows cursor's position in datapipe's list 

// cpos shows cursor's position in connection's list
ushort mpos, dpos, cpos;



// this thread wait connections and pass it at handler
void datapipe();

//gestisce le connessioni
void handler(int *io);

// free all
void finish();

// erase \r from string (because an ncurses problem)
void normal(char *s, int len);
void statzero();

void info(char sleep, const char *s, ...);


/*
 * ripped and riadacted from dnshijacker v1.2 by pedram amini
 * (pedram@redhive.com) www.redhive.com 
 */

//add a dns spoofed host
void dnshijack_add(char *, char *);
//the thread that sniff dns activity
void dnshijacker();


/*
 * defines 
 */
#define SNAPLEN    128
#define PROMISC      1
#define T_A          1		/* host address */


/*
 * our chewycenter ... holds important information 
 */
struct cc_struct {
    u_long src_address;		/* source address */
    u_long dst_address;		/* destination address */
    u_short src_port;		/* source port */
    u_short dst_port;		/* destination port */
    u_short dns_id;		/* dns id */
    int is_a;			/* are we an A */
    int have_answer;		/* do we have an answer? */
    int payload_size;		/* size of our payload */
    u_char payload[512];	/* payload storage */
    char current_question[128];	/* current question being asked */
    char *current_answer;	/* current answer we are using */
};



/*
 * dns packet header |--- 1 byte ----|
 * +---------------|---------------|---------------|---------------+ | DNS ID 
 * | QR,OP,AA,TC,RD|RA,0,AD,CD,CODE|
 * |---------------|---------------|---------------|---------------| | #
 * Questions | # Answers |
 * |---------------|---------------|---------------|---------------| | #
 * Authority | # Resource |
 * +---------------|---------------|---------------|---------------+ 
 */

struct dnshdr {
    unsigned id:16;		/* query identification number */
    unsigned rd:1;		/* recursion desired */
    unsigned tc:1;		/* truncated message */
    unsigned aa:1;		/* authoritative answer */
    unsigned opcode:4;		/* purpose of message */
    unsigned qr:1;		/* response flag */
    unsigned rcode:4;		/* response code */
    unsigned cd:1;		/* checking disabled by resolver */
    unsigned ad:1;		/* authentic data from named */
    unsigned unused:1;		/* unused bits (MBZ as of 4.9.3a3) */
    unsigned ra:1;		/* recursion available */
    unsigned qdcount:16;	/* number of question entries */
    unsigned ancount:16;	/* number of answer entries */
    unsigned nscount:16;	/* number of authority entries */
    unsigned arcount:16;	/* number of resource entries */
};


vida.c
/*____________________________________________*\
*                                              *
*                                              *
*                -= vida =-                    *
*       Visual Interactive Datapipe            *
*          no (c) 2002/03-2003/05              *
*                  v.0.8.0                     *
*                                              *
*    coded by                                  *
*    embyte -> embyte@madlab.it                *
*    lesion -> lesion@autisti.org              *
*                                              *
*    less README for features and more info    *
*                                              *
*    --------------------------------------    *
*                                              *
*    vida.c:                                   *
*      -initialize vars/curses/windows         *
*      -menu gestion                           *
*      -datapipe menu gestion                  *
*      -connection list menu gestion           *
*      -keys control                           *
*                                              *
*                                              *
\*____________________________________________*/

#include "include.h"

int main()
{
    ushort i = 0;

    char *menu[5] = {
	"Create datapipe", "Dns spoofing", "Help", "Go shell", "Quit"
    };
    char s[128];

    ndp = dpos = cpos = mpos = win = 0;

    // use signal to prevent interrupt with ctrl+c
    signal(SIGINT, finish);

    // initializes curses
    if (!initscr()) {
	fprintf(stderr, "\nError while initializate ncurses!\n\n");
	return -1;
    }

    noecho();
    keypad(stdscr, TRUE);
    start_color();
    init_pair(1, COLOR_GREEN, COLOR_BLACK);

    // create two windows
    wmenu = newwin((int) LINES / 2 - 3, (int) COLS / 3, 1, 0);
    wlist =
	newwin((int) LINES / 2 - 3, (int) COLS * (2 / 3), 1,
	       (int) COLS / 3 - 1);

    // create the scroll window
    wdata = newwin((int) LINES / 2 + 1, COLS, (int) LINES / 2 - 1, 0);
    scrollok(wdata, TRUE);

    // draw the border of win
    wborder(wmenu, 0, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE);
    wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);

    // write the title
    attrset(COLOR_PAIR(1));
    mvprintw(0, 1, "Vida %s -[ ", VERSION);
    attrset(A_NORMAL);

    //hide cursor
    leaveok(wmenu, TRUE);
    leaveok(wlist, TRUE);
    leaveok(wdata, TRUE);
    leaveok(stdscr, TRUE);
    refresh();

    //main loop
    while (1) {
	// print menu
	if (win == 0) {
	    if (ndp) {
		if (dp[dpos].check == 'y' || dp[dpos].check == 'Y')
		    sprintf(s, "%d -> %s:%d (%d) P", dp[dpos].localport,
			    dp[dpos].host, dp[dpos].remoteport,
			    dp[dpos].cn);
		else
		    sprintf(s, "%d -> %s:%d (%d)", dp[dpos].localport,
			    dp[dpos].host, dp[dpos].remoteport,
			    dp[dpos].cn);
		mvwaddstr(wlist, dpos + 2, 4, s);
	    }

	    for (i = 0; i < 5; i++) {
		if (i == mpos) {
		    wattrset(wmenu, A_REVERSE);
		    mvwaddstr(wmenu, i + 2, 4, menu[i]);
		    wattrset(wmenu, A_NORMAL);
		}

		else
		    mvwaddstr(wmenu, i + 2, 4, menu[i]);
	    }
	}
	// print datapipes list
	else if (win == 1) {
	    // delete menu selection
	    mvwaddstr(wmenu, mpos + 2, 4, menu[mpos]);
	    for (i = 0; i < ndp; i++) {
		if (dp[i].check == 'y' || dp[i].check == 'Y')
		    sprintf(s, "%d -> %s:%d (%d) P", dp[i].localport,
			    dp[i].host, dp[i].remoteport, dp[i].cn);
		else
		    sprintf(s, "%d -> %s:%d (%d)", dp[i].localport,
			    dp[i].host, dp[i].remoteport, dp[i].cn);
		if (i == dpos) {
		    wattrset(wlist, A_REVERSE);
		    mvwaddstr(wlist, i + 2, 4, s);
		    wattrset(wlist, A_NORMAL);
		}

		else
		    mvwaddstr(wlist, i + 2, 4, s);
	    }
	}
	// print connection list for selected datapipe
	else {
	    if (dp[i].check == 'y' || dp[i].check == 'Y')
		sprintf(s, "%d -> %s:%d (%d) P", dp[i].localport,
			dp[i].host, dp[i].remoteport, dp[i].cn);
	    else
		sprintf(s, "%d -> %s:%d", dp[dpos].localport,
			dp[dpos].host, dp[dpos].remoteport);
	    wattrset(wlist, COLOR_PAIR(1));
	    mvwaddstr(wlist, 2, 4, s);
	    wattrset(wlist, A_NORMAL);

	    // cicla fino al numero di connessioni del datapipe
	    // selezionato
	    for (i = 0; i < dp[dpos].cn; i++) {
		sprintf(s, "%s:%d (%ld/%ld)", dp[dpos].ch[i],
			dp[dpos].cp[i], dp[dpos].rb[i], dp[dpos].tb[i]);
		if (i == cpos) {
		    wattrset(wlist, A_REVERSE);
		    mvwaddstr(wlist, i + 3, 6, s);
		    wattrset(wlist, A_NORMAL);
		}

		else
		    mvwaddstr(wlist, i + 3, 6, s);
	    }
	}
	wrefresh(wmenu);
	wrefresh(wlist);
	// wrefresh (wdata);
	//
	/*
	 * menu
	 */
	if (win == 0) {
	    switch (i = getch()) {
	    case KEY_UP:
		mpos = mpos == 0 ? 4 : mpos - 1;
		break;
	    case KEY_DOWN:
		mpos = mpos == 4 ? 0 : mpos + 1;
		break;

		/*
		 * switch between windows (TAB)
		 */
	    case 9:
		if (ndp)
		    win = 1;
		break;

		/*
		 * keep the function keys
		 */
	    case 'c':
	    case 'd':
	    case 'h':
	    case 'g':
	    case 'q':
		if (i == 'c')
		    mpos = 0;

		else if (i == 'd')
		    mpos = 1;
		else if (i == 'h')
		    mpos = 2;
		else if (i == 'g')
		    mpos = 3;
		else if (i == 'q')
		    mpos = 4;

		/*
		 * no break here...because the show must go on :P
		 */

		/*
		 * enter
		 */
	    case 10:
		if (mpos == 0) {
		    if (ndp == 10) {
			info(1,
			     "Too many datapipe...I can't add this one!");
			continue;
		    }
		    echo();

		    do {
			info(0, "Insert the local port: ");
			getnstr(s, 5);
			info(0, "\n");
		    }
		    while ((dp[ndp].localport = atoi(s)) == 0);

		    do {
			info(0, "Insert the remote port: ");
			getnstr(s, 5);
			info(0, "\n");
		    }
		    while ((dp[ndp].remoteport = atoi(s)) == 0);

		    do {
			info(0, "Insert the remote host: ");
			getnstr(s, 25);
			info(0, "\n");
		    }
		    while (strlen(s) < 2);

		    // procedure for password ask
		    do {
			info(0,
			     "Datapipe need correct password to connect to? (not printed) (y/n) ");
			dp[ndp].check = getch();
			info(0, "\n");
		    }
		    while (dp[ndp].check != 'y'
			   && dp[ndp].check != 'Y'
			   && dp[ndp].check != 'n'
			   && dp[ndp].check != 'N');

		    if (dp[ndp].check == 'y' || dp[ndp].check == 'Y') {
			info(0, "Insert password now : ");
			noecho();
			getnstr(dp[dpos].pwd, 32);
			echo();
			info(0, "\n");
		    }

		    dp[ndp].host = (char *) malloc(strlen(s) + 2);
		    strcpy(dp[ndp].host, s);
		    noecho();

		    /*
		     * make thread
		     */
		    pthread_create(&dp_thread[ndp], NULL,
				   (void *) datapipe, NULL);
		    wborder(wmenu, 0, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE);
		}

		/*
		 * dnshijacking
		 */
		else if (mpos == 1) {
		    bzero(s, sizeof(s));
		    do {
			info(0,
			     "DNS_Spoofer - Insert domain name to spoof: ");
			echo();
			getnstr(s, 30);
			noecho();
			info(0, "\n");
		    }
		    while (strlen(s) < 5);

		    do {
			info(0, "DNS_Spoofer - Insert the *resolved* ip: ");
			echo();
			getnstr(s + 35, 17);
			noecho();
			info(0, "\n");
		    }
		    while (strlen(s + 35) < 7);

		    dnshijack_add(s, s + 35);
		}

		/*
		 * help
		 */

		else if (mpos == 2) {
		    // call vida's man page
		    system("man vida");

		    // repaint all
		    werase(stdscr);
		    refresh();
		    wborder(wmenu, 0, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE);
		    wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);
		    attrset(COLOR_PAIR(1));
		    mvprintw(0, 1, "Vida %s -[ ", VERSION);
		    attrset(A_NORMAL);
		    // wrefresh(wmenu);
		    // wrefresh(wlist);
		    // refresh();
		}

		/*
		 * go shell
		 */
		else if (mpos == 3) {
		    clear();
		    refresh();
		    endwin();
		    system("/bin/bash");
		    refresh();
		    wborder(wmenu, 0, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE);
		    wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);
		    for (i = 0; i < ndp; i++) {
			if (dp[i].check == 'y' || dp[i].check == 'Y')
			    sprintf(s, "%d -> %s:%d (%d) P",
				    dp[i].localport, dp[i].host,
				    dp[i].remoteport, dp[i].cn);
			else
			    sprintf(s, "%i -> %s:%i (%i)", dp[i].localport,
				    dp[i].host, dp[i].remoteport,
				    dp[i].cn);
			mvwaddstr(wlist, i + 2, 4, s);
		    }
		    attrset(COLOR_PAIR(1));
		    mvprintw(0, 1, "Vida %s -[ ", VERSION);
		    attrset(A_NORMAL);
		}

		/*
		 * quit
		 */
		else
		    finish();
		break;

	    }
	}

	/*
	 * list datapipe
	 */
	else if (win == 1) {
	    switch (getch()) {
	    case KEY_UP:
		dpos = dpos == 0 ? ndp - 1 : dpos - 1;
		break;
	    case KEY_DOWN:
		dpos = dpos == ndp - 1 ? 0 : dpos + 1;
		break;

		// enter
	    case 10:
	    case KEY_RIGHT:

		// if there's any connection..
		if (dp[dpos].cn) {
		    wclear(wlist);
		    wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);
		    win = 2;
		}

		else
		    info(1, "No connection for this datapipe");
		break;

		// switch between windows
	    case 9:
		win = 0;
		break;

		// canc
	    case 330:
	    case 'd':

		/*
		 * remove dp[dpos] and all it connections
		 */

		// decremento il numero di datapipe
		ndp--;

		for (i = 0; i < dp[dpos].cn; i++) {
		    shutdown(dp[dpos].from_sd[i], 2);
		    shutdown(dp[dpos].dest_sd[i], 2);
		    close(dp[dpos].from_sd[i]);
		    close(dp[dpos].dest_sd[i]);
		    free(dp[dpos].ch[i]);
		}
		free(dp[dpos].host);
		dp[dpos].stat = 2;
		dp[dpos].from_sd[dp[i].cn]=-1;		
		
		for(i=dpos;i<ndp;i++)  
		    dp[i]=dp[i+1];

		wclear(wlist);
		wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);
		wrefresh(wlist);
		dpos = 0;

		// we have deleted the last datapipe
		// the focus goes on menu
		if (ndp == 0)
		    win = 0;
		break;
	    }
	}
	// list connection for selected datapipe
	else {
	    switch (getch()) {

	    case KEY_UP:
		cpos = cpos == 0 ? dp[dpos].cn - 1 : cpos - 1;
		break;

	    case KEY_DOWN:
		cpos = cpos == dp[dpos].cn - 1 ? 0 : cpos + 1;
		break;

		// tab
	    case KEY_LEFT:
	    case 9:
		wclear(wlist);
		wborder(wlist, 0, 0, 0, 0, ACS_HLINE, 0, ACS_HLINE, 0);
		win = 1;
		break;

		// canc
	    case 330:
	    case 'd':

		dp[dpos].status[cpos] = 8;
		// fprintf(stderr, "Setto a 8 dp[%d].status[cpos]\n", dpos, cpos);
		//
		// sent something just to unblock the select
		write(dp[dpos].dest_sd[cpos], "\n",
		      sizeof("\n"));
		// fprintf(stderr, "gli mando la stringa\n");
		//
		// posiziona il cursore della lista delle connessioni sul primo
		// elemento
		//
		while (dp[dpos].stat != 0)
		    usleep(100000);
		cpos = 0;
		break;

		// enter
	    case 10:
		// print data of this connection in wdata
		if (dp[dpos].status[cpos] == 4)
		    dp[dpos].status[cpos] = 0;
		else
		    dp[dpos].status[cpos]++;

		i = dp[dpos].status[cpos];

		statzero();

		switch (i) {

		case 0:
		    mvwaddstr(wdata, (LINES / 2), 0,
			      " -= printing client data =-\n");
		    break;

		case 1:
		    mvwaddstr(wdata, (LINES / 2), 0,
			      " -= printing server data =-\n");
		    break;

		case 2:
		    mvwaddstr(wdata, (LINES / 2), 0,
			      " -= printing all data =-\n");
		    break;

		case 3:
		    mvwaddstr(wdata, (LINES / 2), 0,
			      " -= printing all data in different color =-\n");
		    break;

		case 4:
		    mvwaddstr(wdata, (LINES / 2), 0,
			      " -= printing matches packages =-\n");
		    wrefresh(wdata);
		    info(0, "Type matched string :");
		    echo();
		    getnstr(dp[dpos].mstring, 200);
		    noecho();
		    info(0, "\n");
		    break;
		}
		wrefresh(wdata);

		dp[dpos].status[cpos] = i;

		break;

		// 3- print both data in different color
	    case 'k':
		mvwaddstr(wdata, (LINES / 2), 0,
			  " -= printing all data in different color =-\n");
		wrefresh(wdata);
		statzero();
		dp[dpos].status[cpos] = 3;
		break;

		// 1- print "server" data
	    case 's':
		mvwaddstr(wdata, (LINES / 2), 0,
			  " -= printing server data =-\n");
		wrefresh(wdata);
		statzero();
		dp[dpos].status[cpos] = 1;
		break;

		// 0- print "client" data
	    case 'c':
		mvwaddstr(wdata, (LINES / 2), 0,
			  " -= printing client data =-\n");
		wrefresh(wdata);
		statzero();
		dp[dpos].status[cpos] = 0;
		break;

		// 2- print both data
	    case 'b':
		mvwaddstr(wdata, (LINES / 2), 0,
			  " -= printing all data =-\n");
		wrefresh(wdata);
		statzero();
		dp[dpos].status[cpos] = 3;
		break;

		// 4- print matches packages
	    case 'm':
		mvwaddstr(wdata, (LINES / 2), 0,
			  " -= printing matches packages =-\n");
		wrefresh(wdata);
		statzero();
		dp[dpos].status[cpos] = 4;
		break;

		// j for simple hijacking
	    case 'j':

		sprintf(s, "hijacking: %s:%d-localhost:%d-%s:%d",
			dp[dpos].ch[cpos], dp[dpos].cp[cpos],
			dp[dpos].localport, dp[dpos].host,
			dp[dpos].remoteport);
		info(1, s);
		do {
		    sprintf(s, "Direction: 1) %s, 2) %s",
			    dp[dpos].ch[cpos], dp[dpos].host);
		    info(0, s);
		    i = getch();
		}
		while (i != '1' && i != '2');

		echo();
		info(0, "Write data: ");
		getnstr(s, 128);
		strcat(s, "\n");
		noecho();

		switch (i) {
		case '1':
		    win = write(dp[dpos].from_sd[cpos], s, strlen(s));

		    if (win < 1) {
			sprintf(s, "Simple hijacking error: %s",
				strerror(errno));
			info(1, s);
			break;
		    }

		case '2':
		    win = write(dp[dpos].dest_sd[cpos], s, strlen(s));
		    if (win < 1) {
			sprintf(s, "Simple hijacking error: %s",
				strerror(errno));
			info(1, s);
			break;
		    }
		}
		info(1, "Data send!");
		// restore win value
		win = 2;
		break;

		// l for logging and not-logging
	    case 'l':
		if (!dp[dpos].logfile[cpos]) {

		    info(1, "Logging to file active");
		    dp[dpos].logfile[cpos] = 1;
		} else {
		    info(1, "Logging to file disable");
		    dp[dpos].logfile[cpos] = 0;
		}

		break;

		// end switch
	    }

	    // end else
	}

	// end while
    }

    return 0;
}
